aws_topology/stackstate_checks/aws_topology/cloudtrail.py

```diff
@@ -0,0 +1,126 @@
+import json
+import dateutil.parser
+import pytz
+from datetime import datetime
+import gzip
+import io
+import botocore
+from six import string_types, PY2
+
+
+try:
+    JSONParseException = json.decoder.JSONDecodeError
+except AttributeError:  # Python 2
+    JSONParseException = ValueError
+
+
+class CloudtrailCollector(object):
+    def __init__(self, bucket_name, account_id, session, agent):
+        self.bucket_name = bucket_name
+        self.session = session
+        self.agent = agent
+        self.account_id = account_id
+
+    def get_messages(self, not_before):
+        try:
+            # try s3
+            return self._get_messages_from_s3(not_before)
+        except Exception:
+            # try loopup_events
+            self.agent.warning("Falling back to slower Cloudtrail lookup_events")
+            return self._get_messages_from_cloudtrail(not_before)
+
+    def _get_bucket_name(self):
+        if self.bucket_name:
+            return self.bucket_name
+        else:
+            return "stackstate-logs-{}".format(self.account_id)
+
+    def _get_messages_from_s3(self, not_before):
+        client = self.session.client("s3")
+        region = client.meta.region_name
+        bucket_name = self._get_bucket_name()
+        to_delete = []
+        files_to_handle = []
+        # s3:ListBucket
+        # NoSuchBucket
+        for pg in client.get_paginator("list_objects_v2").paginate(
+            Bucket=bucket_name, Prefix="AWSLogs/{act}/EventBridge/{rgn}/".format(act=self.account_id, rgn=region)
+        ):
+            for itm in pg.get("Contents") or []:
+                pts = []
+                try:
+                    pts = [int(i) for i in itm["Key"][-59 : -59 + 19].split("-")]
+                except ValueError:
+                    pass
+                if len(pts) == 6:
+                    dt = datetime(pts[0], pts[1], pts[2], pts[3], pts[4], pts[5]).replace(tzinfo=pytz.utc)
```

Comment:
Regex with match patterns might simplify this a bit, depending on how complex the regex needs to be?